### PR TITLE
After entity start tracking event (1.19.3)

### PR DESCRIPTION
**Pull request by williambl**
# Motivation

The current entity tracking event, `EntityTrackingEvents#START_TRACKING`, is fired before the client is aware of the entity. Therefore, you can't actually send any packets to the player that reference the entity.

# Content

This PR:
 - Adds `EntityTrackingEvents#AFTER_START_TRACKING`.
 - Copies `#START_TRACKING` -> `#BEFORE_START_TRACKING`, with `#START_TRACKING` marked as deprecated.
 - Adds javadoc explaining the differences between the methods.
 - Adds a testmod for entity tracking events.

### CODE REVIEW

**Review by Leo40Git — APPROVED** [review]

File: library/core/networking/src/main/java/org/quiltmc/qsl/networking/api/EntityTrackingEvents.java [hunk 1]
```diff
@@ -30,13 +30,43 @@ public final class EntityTrackingEvents {
...
 
+	/**
+	 * An event that is called before player starts tracking an entity.
+	 * Typically, this occurs when an entity enters a client's view distance.
+	 * This event is called before the player's client is sent the entity's {@link Entity#createSpawnPacket() spawn packet}.
+	 *
+	 * @apiNote Since the client will not know about the entity at this point, you probably don't want to send packets
+	 * referencing the entity here. Do that in {@link #AFTER_START_TRACKING} instead.
+	 *
+	 * @deprecated Use {@link #BEFORE_START_TRACKING} instead. It's the same event with a more descriptive name.
```
Comment (Leo40Git) [inline library/core/networking/src/main/java/org/quiltmc/qsl/networking/api/EntityTrackingEvents.java]:
```suggestion
	 * @deprecated Renamed to {@link #BEFORE_START_TRACKING}.
```
...maybe? I feel like the original sentence is too wordy.

Comment (williambl) [reply library/core/networking/src/main/java/org/quiltmc/qsl/networking/api/EntityTrackingEvents.java]: Yeah, that's much better.

File: library/core/networking/src/testmod/java/org/quiltmc/qsl/networking/test/trackingevents/NetworkingEntityTrackingEventsTest.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
+/*
+ * Copyright 2022 QuiltMC
```
Comment (EnnuiL) [inline library/core/networking/src/testmod/java/org/quiltmc/qsl/networking/test/trackingevents/NetworkingEntityTrackingEventsTest.java]: It's 2023!
```suggestion
 * Copyright 2023 QuiltMC
```

**Review by EnnuiL — APPROVED** [review]
Looks good to me! Just one nitpick:

File: library/core/networking/src/testmod/java/org/quiltmc/qsl/networking/test/trackingevents/NetworkingEntityTrackingEventsTest.java (hunk 2 above)
Comment (williambl) [reply library/core/networking/src/testmod/java/org/quiltmc/qsl/networking/test/trackingevents/NetworkingEntityTrackingEventsTest.java]: meme - 'there i go, still writing 'B.C.' on my checks'

**Review by OroArmor — APPROVED** [review]

**Review by EnnuiL — CHANGES_REQUESTED** [review]
wait heck, I do have one objection:
Shouldn't the StartTracking event also be renamed to BeforeStartTracking?

File: library/core/networking/src/testmod/java/org/quiltmc/qsl/networking/test/trackingevents/NetworkingEntityTrackingEventsTest.java [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+import net.minecraft.server.network.ServerPlayerEntity;
+import net.minecraft.text.Text;
+
+import org.quiltmc.qsl.networking.api.EntityTrackingEvents;
+
+public final class NetworkingEntityTrackingEventsTest implements EntityTrackingEvents.AfterStartTracking, EntityTrackingEvents.StartTracking, EntityTrackingEvents.StopTracking {
+	// Sends a message to the player when they're about to start tracking an entity.
+	@Override
+	public void onStartTracking(Entity trackedEntity, ServerPlayerEntity player) {
+		player.sendMessage(Text.literal("About to be tracking ").append(trackedEntity.getDisplayName()), false);
```
Comment (EnnuiL) [inline library/core/networking/src/testmod/java/org/quiltmc/qsl/networking/test/trackingevents/NetworkingEntityTrackingEventsTest.java]: aaaaaaaaaaa this causes log *and* chat spam ;-;
i wonder if something more ephemeral could be achieved instead, or maybe there should be like, a toggle for this

**Review by EnnuiL — CHANGES_REQUESTED** [review]

File: library/core/networking/src/testmod/java/org/quiltmc/qsl/networking/test/trackingevents/NetworkingEntityTrackingEventsTest.java (hunk 3 above)
Comment (williambl) [reply library/core/networking/src/testmod/java/org/quiltmc/qsl/networking/test/trackingevents/NetworkingEntityTrackingEventsTest.java]: Made this only trigger for armour stands.

**Comment by williambl** [comment]
> wait heck, I do have one objection: Shouldn't the StartTracking event also be renamed to BeforeStartTracking?

Renaming the interface would break mods currently using the event.

**Comment by Pyrofab** [comment]
Maybe you could add the `BeforeStartTracking` interface and make `StartTracking` inherit from it (and deprecate the latter)? This way, if I'm not mistaken:
- binary compatibility is preserved (`StartTracking` implementations are valid `BeforeStartTracking` implementations too)
- recompiled lambdas are automatically converted to `BeforeStartTracking` implementations
- explicit implementations are marked as deprecated

**Comment by EnnuiL** [comment]
Exempting an important note from break: Do note that we have 1.19.4 incoming, and that's a chance for breaking changes

**Review by EnnuiL — APPROVED** [review]
I hope to get this merged just in time for 1.19.4, so we can remove the deprecated redirect :P

**Review by LambdAurora — APPROVED** [review]

**Comment by EnnuiL** [comment]
With tests done? This PR has entered FCP for 5 days! If needed, this can be fast-tracked in order to guarantee the necessary refactoring is done on 1.19.4
